### PR TITLE
Fixes isValidClosureName to handle generic types correctly.

### PR DESCRIPTION
**Pull request by asifmujteba**
Fixes krzysztofzablocki/Sourcery#844

### CODE REVIEW

File: SourceryTests/Models/TypeNameSpec.swift [hunk 1]
```diff
@@ -111,6 +111,9 @@ class TypeNameSpec: QuickSpec {
                     expect(TypeName("() -> (Int, Int)").isClosure).to(beTrue())
                     expect(TypeName("() -> (Int) -> (Int)").isClosure).to(beTrue())
                     expect(TypeName("((Int) -> (Int)) -> ()").isClosure).to(beTrue())
+                    expect(TypeName("(Foo<String>) -> Bool)").isClosure).to(beTrue())
+                    expect(TypeName("(Int) -> Foo<Bool>").isClosure).to(beTrue())
+                    expect(TypeName("(Foo<String>) -> Foo<Bool>)").isClosure).to(beTrue())
```
Comment (ilyapuchka) [inline SourceryTests/Models/TypeNameSpec.swift:116]: nit: this does not seem to be a valid type actually, the last ) is extra, did you mean it as a single value tuple?

File: SourceryTests/Models/TypeNameSpec.swift [hunk 2]
```diff
@@ -130,6 +133,16 @@ class TypeNameSpec: QuickSpec {
                 }
             }
 
+            context("given closure type inside generic type") {
+                it("reports closure correctly") {
+                    expect(TypeName("Foo<() -> ()>").isClosure).to(beFalse())
+                    expect(TypeName("Foo<(String) -> Bool>").isClosure).to(beFalse())
+                    expect(TypeName("Foo<(String) -> Bool?>").isClosure).to(beFalse())
+                    expect(TypeName("Foo<(Bar<String>) -> Bool)>").isClosure).to(beFalse())
+                    expect(TypeName("Foo<(Bar<String>) -> Bar<Bool>)>").isClosure).to(beFalse())
```
Comment (ilyapuchka) [inline SourceryTests/Models/TypeNameSpec.swift:142]: ditto, brackets are not balanced, not sure if that's on purpose?

File: SourceryTests/Models/TypeNameSpec.swift (hunk 1 above)
Comment (asifmujteba) [reply SourceryTests/Models/TypeNameSpec.swift]: thanks for pointing out, it was a typo, fixed now.

File: SourceryTests/Models/TypeNameSpec.swift (hunk 2 above)
Comment (asifmujteba) [reply SourceryTests/Models/TypeNameSpec.swift]: thanks for pointing out, it was a typo, fixed now.

**Comment by jpsim** [comment]
This change introduced a crash in my use of Sourcery: #856

**Pull request merged**